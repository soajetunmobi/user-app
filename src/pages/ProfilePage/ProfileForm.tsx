import React, { FunctionComponent, useState } from 'react'
import { Button, createStyles, Grid, TextField } from '@material-ui/core'
import { Field, Form, Formik, FormikProps } from 'formik'
import { makeStyles } from '@material-ui/core/styles'
import { LoadingIndicator } from 'components/shared/LoadingIndicator'
import { profileFormValidation } from 'validation'
import { useUserContext } from 'providers'

interface IProfileForm {
  firstname: string
  othernames: string
  street: string
  town: string
  county: string
  postcode: string
  mobile: string
  email: string
  company: string
  preferences: string[]
}

const formStatusProps: IFormStatusProps = {
  success: {
    message: 'Profile update successful.',
    type: 'success'
  },
  fail: {
    message: 'Profile update failed',
    type: 'error'
  },
  error: {
    message: 'Something went wrong. Please try again.',
    type: 'error'
  }
}

const useStyles = makeStyles(() =>
  createStyles({
    root: {
      maxWidth: '700px',
      display: 'block',
      margin: '0 auto'
    },
    textField: {
      '& > *': {
        width: '100%'
      }
    },
    submitButton: {
      marginTop: '24px'
    },
    title: { textAlign: 'center' },
    successMessage: { color: 'green' },
    errorMessage: { color: 'red' }
  })
)

interface ProfileFormProps {
  user: User
  cancelEdit: () => void
}

export const ProfileForm: FunctionComponent<ProfileFormProps> = ({ user, cancelEdit }) => {
  const classes = useStyles()
  const { setUser } = useUserContext()
  const [displayFormStatus, setDisplayFormStatus] = useState(false)
  const [formStatus, setFormStatus] = useState<IFormStatus>({
    message: '',
    type: ''
  })

  const initialValue = {
    firstname: user.first_name,
    othernames: user.other_names,
    street: user.address.street,
    town: user.address.town,
    county: user.address.county,
    postcode: user.address.postcode,
    mobile: user.mobile,
    email: user.email,
    company: user.company,
    preferences: [...user.preferences.contact]
  }

  const saveProfileUpdate = async (data: IProfileForm) => {
    const updatedUser = {
      id: user.id,
      first_name: data.firstname,
      other_names: data.othernames,
      address: {
        street: data.street,
        town: data.town,
        county: data.county,
        postcode: data.postcode
      },
      mobile: data.mobile,
      email: data.email,
      company: data.company,
      preferences: {
        contact: [...data.preferences]
      }
    }
    try {
      setFormStatus(formStatusProps.success)
      setUser(updatedUser)
      cancelEdit()
    } catch (err) {
      setFormStatus(formStatusProps.fail)
    }

    setDisplayFormStatus(true)
  }

  if (!user) {
    return <LoadingIndicator />
  }

  return (
    <div className={classes.root} data-testid='profile-edit'>
      <Formik
        initialValues={initialValue}
        onSubmit={async (values: IProfileForm, actions) => {
          await saveProfileUpdate(values)
          setTimeout(() => {
            actions.setSubmitting(false)
          })
        }}
        validationSchema={profileFormValidation}
      >
        {(props: FormikProps<IProfileForm>) => {
          const { values, touched, errors, handleBlur, handleChange, isSubmitting } = props
          return (
            <Form>
              <h1 className={classes.title}>Profile - Edit</h1>
              <Grid container justify='space-around' direction='row'>
                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
                  <TextField
                    name='firstname'
                    id='firstname'
                    label='First name'
                    value={values.firstname}
                    type='text'
                    helperText={
                      errors.firstname && touched.firstname ? errors.firstname : 'Enter firstname'
                    }
                    error={errors.firstname && touched.firstname ? true : false}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                </Grid>
                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
                  <TextField
                    name='othernames'
                    id='othernames'
                    label='Other names'
                    value={values.othernames}
                    type='text'
                    helperText={
                      errors.othernames && touched.othernames
                        ? errors.othernames
                        : 'Enter other names'
                    }
                    error={errors.othernames && touched.othernames ? true : false}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                </Grid>
                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
                  <TextField
                    name='street'
                    id='street'
                    label='Street'
                    value={values.street}
                    type='text'
                    helperText={errors.street && touched.street ? errors.street : 'Enter street'}
                    error={errors.street && touched.street ? true : false}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                </Grid>
                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
                  <TextField
                    name='town'
                    id='town'
                    label='Town'
                    value={values.town}
                    type='text'
                    helperText={errors.town && touched.town ? errors.town : 'Enter town'}
                    error={errors.town && touched.town ? true : false}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                </Grid>
                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
                  <TextField
                    name='county'
                    id='county'
                    label='County'
                    value={values.county}
                    type='text'
                    helperText={errors.county && touched.county ? errors.county : 'Enter county'}
                    error={errors.county && touched.county ? true : false}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                </Grid>
                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
                  <TextField
                    name='postcode'
                    id='postcode'
                    label='Postcode'
                    value={values.postcode}
                    type='text'
                    helperText={
                      errors.postcode && touched.postcode ? errors.postcode : 'Enter postcode'
                    }
                    error={errors.postcode && touched.postcode ? true : false}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                </Grid>
                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
                  <TextField
                    name='mobile'
                    id='mobile'
                    label='Mobile'
                    value={values.mobile}
                    type='text'
                    helperText={errors.mobile && touched.mobile ? errors.mobile : 'Enter mobile'}
                    error={errors.mobile && touched.mobile ? true : false}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                </Grid>
                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
                  <TextField
                    name='email'
                    id='email'
                    label='Email'
                    value={values.email}
                    disabled={true}
                    type='email'
                    helperText={errors.email && touched.email ? errors.email : 'Enter email'}
                    error={errors.email && touched.email ? true : false}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                </Grid>
                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
                  <TextField
                    name='company'
                    id='company'
                    label='Company'
                    value={values.company}
                    type='text'
                    helperText={
                      errors.company && touched.company ? errors.company : 'Enter company'
                    }
                    error={errors.company && touched.company ? true : false}
                    onChange={handleChange}
                    onBlur={handleBlur}
                  />
                </Grid>

                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
                  <div role='group' aria-labelledby='checkbox-group'>
                    <label>
                      <Field type='checkbox' name='preferences' value='sms' />
                      SMS
                    </label>
                    <label>
                      <Field type='checkbox' name='preferences' value='mail' />
                      Mail
                    </label>
                    <label>
                      <Field type='checkbox' name='preferences' value='email' />
                      Email
                    </label>
                  </div>
                </Grid>
                <Grid item lg={10} md={10} sm={10} xs={10} className={classes.submitButton}>
                  <Button
                    type='reset'
                    variant='outlined'
                    color='secondary'
                    data-testid='profile-cancel-button'
                    onClick={cancelEdit}
                  >
                    Cancel
                  </Button>
                  <Button
                    type='submit'
                    variant='contained'
                    color='secondary'
                    disabled={isSubmitting}
                  >
                    Save
                  </Button>
                  {displayFormStatus && (
                    <div className='formStatus'>
                      {formStatus.type === 'error' ? (
                        <p className={classes.errorMessage}>{formStatus.message}</p>
                      ) : formStatus.type === 'success' ? (
                        <p className={classes.successMessage}>{formStatus.message}</p>
                      ) : null}
                    </div>
                  )}
                </Grid>
              </Grid>
            </Form>
          )
        }}
      </Formik>
    </div>
  )
}
