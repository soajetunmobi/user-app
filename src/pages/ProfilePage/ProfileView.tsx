import React, { FunctionComponent } from 'react'
import { Button, createStyles, Grid } from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import { LoadingIndicator } from 'components/shared/LoadingIndicator'

const useStyles = makeStyles(() =>
  createStyles({
    root: {
      maxWidth: '450px',
      display: 'block',
      margin: '0 auto'
    },
    textField: {
      '& > *': {
        width: '100%'
      }
    },
    submitButton: {
      marginTop: '24px'
    },
    title: { textAlign: 'center' },
    successMessage: { color: 'green' },
    errorMessage: { color: 'red' }
  })
)

interface ProfileViewProps {
  user: User
  cancelView: () => void
  editContent: () => void
}

export const ProfileView: FunctionComponent<ProfileViewProps> = ({
  user,
  cancelView,
  editContent
}) => {
  const classes = useStyles()

  const renderAddress = () => {
    const { address } = user
    return (
      <div>
        <div>Street: {address.street}</div>
        <div>Town: {address.town}</div>
        <div>County: {address.county}</div>
        <div>Postcode: {address.postcode}</div>
      </div>
    )
  }
  const renderPreferences = () => {
    const {
      preferences: { contact }
    } = user
    return (
      <div>
        {contact.map((choice: preference) => {
          return <span key={choice}>{choice}</span>
        })}
      </div>
    )
  }

  if (!user) {
    return <LoadingIndicator />
  }

  return (
    <div className={classes.root} data-testid='profile-view'>
      <h1 className={classes.title}>Profile - View</h1>
      <Grid container justify='space-around' direction='row'>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Firstname:</span>
          <span data-testid='profile-firstname'>{user.first_name}</span>
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Other names:</span>
          <span data-testid='profile-othernames'>{user.other_names}</span>
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Address:</span>
          {renderAddress()}
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Mobile:</span>
          <span>{user.mobile}</span>
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Email:</span>
          <span>{user.mobile}</span>
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Company:</span>
          <span>{user.company}</span>
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.textField}>
          <span>Preferences:</span>
          {renderPreferences()}
        </Grid>
        <Grid item lg={10} md={10} sm={10} xs={10} className={classes.submitButton}>
          <Button
            type='submit'
            variant='outlined'
            color='secondary'
            data-testid='profile-cancel-button'
            onClick={cancelView}
          >
            Cancel
          </Button>
          <Button
            type='submit'
            variant='contained'
            color='secondary'
            data-testid='profile-edit-button'
            onClick={editContent}
          >
            Edit
          </Button>
        </Grid>
      </Grid>
    </div>
  )
}
