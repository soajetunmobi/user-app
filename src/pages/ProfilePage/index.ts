export * from './ProfilePage'
export * from './ProfileView'
export * from './ProfileForm'
