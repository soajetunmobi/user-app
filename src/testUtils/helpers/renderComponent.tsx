import React, { ReactElement, ReactNode } from 'react'
import { render } from '@testing-library/react'
import { UserContext } from 'providers'

const renderComponent = (component: ReactElement, user: User): ReactNode => {
  return render(
    <UserContext.Provider value={{ user, setUser: user => user }}>{component}</UserContext.Provider>
  )
}

export default renderComponent
