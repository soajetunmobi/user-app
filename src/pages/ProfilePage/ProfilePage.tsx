import React, { FunctionComponent, useState } from 'react'
import { useUserContext } from 'providers'
import { LoadingIndicator } from 'components/shared/LoadingIndicator'
import { useHistory } from 'react-router-dom'
import { routes } from 'const/routes'
import { ProfileView } from './ProfileView'
import { ProfileForm } from './ProfileForm'

type viewMode = 'view' | 'edit'

export const ProfilePage: FunctionComponent = () => {
  const { user } = useUserContext()
  const history = useHistory()
  const [currentView, setCurrentView] = useState<viewMode>('view')

  const cancelView = () => {
    history.push(routes.home)
  }

  const cancelEdit = () => {
    setCurrentView('view')
  }

  const editContent = () => {
    setCurrentView('edit')
  }

  if (!user) {
    return <LoadingIndicator />
  }

  return (
    <div>
      {currentView === 'view' && (
        <ProfileView user={user} cancelView={cancelView} editContent={editContent} />
      )}
      {currentView === 'edit' && <ProfileForm user={user} cancelEdit={cancelEdit} />}
    </div>
  )
}
