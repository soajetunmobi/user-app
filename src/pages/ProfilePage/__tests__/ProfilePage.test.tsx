import renderComponent from 'testUtils/helpers/renderComponent'
import { MOCKED_USER } from 'testUtils/mocks/user'
import { fireEvent, screen } from '@testing-library/react'
import React from 'react'
import { ProfilePage } from '../ProfilePage'

describe('ProfilePage', () => {
  it('profile page is rendered', () => {
    expect(renderComponent(<ProfilePage />, MOCKED_USER).asFragment()).toMatchSnapshot()
  })

  describe('Profile views', () => {
    it('should render profile view and not edit', () => {
      renderComponent(<ProfilePage />, MOCKED_USER)
      expect(screen.getByTestId('profile-view')).toBeInTheDocument()
      expect(screen.queryByTestId('/profile-edit/i')).not.toBeInTheDocument()
    })

    it('should render profile edit and not view', () => {
      renderComponent(<ProfilePage />, MOCKED_USER)
      const submit = screen.getByTestId('profile-edit-button')
      fireEvent.click(submit)
      expect(screen.getByTestId('profile-edit')).toBeInTheDocument()
      expect(screen.queryByTestId('/profile-view/i')).not.toBeInTheDocument()
    })

    it('should go back to profile view when I click cancel in edit mode', () => {
      renderComponent(<ProfilePage />, MOCKED_USER)
      const editButton = screen.getByTestId('profile-edit-button')
      fireEvent.click(editButton)
      expect(screen.getByTestId('profile-edit')).toBeInTheDocument()
      const cancelButton = screen.getByTestId('profile-cancel-button')
      fireEvent.click(cancelButton)
      expect(screen.getByTestId('profile-view')).toBeInTheDocument()
    })
  })
})
