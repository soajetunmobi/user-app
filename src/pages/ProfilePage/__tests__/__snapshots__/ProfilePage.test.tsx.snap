// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePage profile page is rendered 1`] = `
<DocumentFragment>
  <div>
    <div
      class="makeStyles-root-1"
      data-testid="profile-view"
    >
      <h1
        class="makeStyles-title-4"
      >
        Profile - View
      </h1>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-space-around"
      >
        <div
          class="MuiGrid-root makeStyles-textField-2 MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-sm-10 MuiGrid-grid-md-10 MuiGrid-grid-lg-10"
        >
          <span>
            Firstname:
          </span>
          <span
            data-testid="profile-firstname"
          >
            Jon
          </span>
        </div>
        <div
          class="MuiGrid-root makeStyles-textField-2 MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-sm-10 MuiGrid-grid-md-10 MuiGrid-grid-lg-10"
        >
          <span>
            Other names:
          </span>
          <span
            data-testid="profile-othernames"
          >
            Williams
          </span>
        </div>
        <div
          class="MuiGrid-root makeStyles-textField-2 MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-sm-10 MuiGrid-grid-md-10 MuiGrid-grid-lg-10"
        >
          <span>
            Address:
          </span>
          <div>
            <div>
              Street: 1 Mill Street
            </div>
            <div>
              Town: Northampton
            </div>
            <div>
              County: Northamponshire
            </div>
            <div>
              Postcode: NU7 JK8
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root makeStyles-textField-2 MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-sm-10 MuiGrid-grid-md-10 MuiGrid-grid-lg-10"
        >
          <span>
            Mobile:
          </span>
          <span>
            08982 92829
          </span>
        </div>
        <div
          class="MuiGrid-root makeStyles-textField-2 MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-sm-10 MuiGrid-grid-md-10 MuiGrid-grid-lg-10"
        >
          <span>
            Email:
          </span>
          <span>
            08982 92829
          </span>
        </div>
        <div
          class="MuiGrid-root makeStyles-textField-2 MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-sm-10 MuiGrid-grid-md-10 MuiGrid-grid-lg-10"
        >
          <span>
            Company:
          </span>
          <span>
            Xerini
          </span>
        </div>
        <div
          class="MuiGrid-root makeStyles-textField-2 MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-sm-10 MuiGrid-grid-md-10 MuiGrid-grid-lg-10"
        >
          <span>
            Preferences:
          </span>
          <div>
            <span>
              mail
            </span>
            <span>
              sms
            </span>
          </div>
        </div>
        <div
          class="MuiGrid-root makeStyles-submitButton-3 MuiGrid-item MuiGrid-grid-xs-10 MuiGrid-grid-sm-10 MuiGrid-grid-md-10 MuiGrid-grid-lg-10"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedSecondary"
            data-testid="profile-cancel-button"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Cancel
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
            data-testid="profile-edit-button"
            tabindex="0"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Edit
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
